sw/device/silicon_creator/lib/sigverify/ecdsa_p256_verify.c

```diff
@@ -0,0 +1,141 @@
+ * kShares[6]` so that xor'ing all shares and `kSigverifySpxSuccess` produces
+ * `kSigverifyFlashExec`, i.e. the value that unlocks flash execution.
+ *
+ * Encoding generated with
+ * $ ./util/design/sparse-fsm-encode.py -d 6 -m 7 -n 32 \
+ *     -s 3118901143 --language=c
+ *
+ * Minimum Hamming distance: 9
+ * Maximum Hamming distance: 25
+ * Minimum Hamming weight: 14
+ * Maximum Hamming weight: 23
+ */
+static const uint32_t kSigverifyShares[kP256SignatureCordWords] = {
+    0xaf28073b, 0x5eb7dcfb, 0x177240b5, 0xa8469df3,
+    0x2e92e9c0, 0x83ed133b, 0x0c9e99f0, 0xc04cd16d,
+};
+
+/**
+ * Checks the encoded message and produces the value to write to the flash_ctrl.
+ *
+ * @param recovered_r Recovered r parameter, little-endian.
+ * @param signature Provided signature, little-endian.
+ * @param[out] flash_exec Value to write to the flash_ctrl EXEC register.
+ * @return Result of the operation.
+ */
+OT_WARN_UNUSED_RESULT
+static rom_error_t sigverify_encoded_message_check(
+    sigverify_ecdsa_p256_buffer_t *recovered_r,
+    const sigverify_ecdsa_p256_buffer_t *signature, uint32_t *flash_exec) {
+  // The algorithm below uses shares, i.e. trivial secret sharing, to check an
+  // encoded message and produce two values: `flash_exec` and `result`.
+  // `flash_exec` is the value to write to the flash_ctrl EXEC register to
+  // unlock flash execution and `result` is the return value. We produce
+  // `result` in addition to `flash_exec` to avoid having the unlock value in
+  // registers or memory just for checking the result of signature verification.
+  // The algorithm consists of two steps:
+  //
+  // 1. First, we xor each word of `recovered_r` with the corresponding expected
+  // value and share (`kSigverifyShares[i]`). At the end of this step,
+  // `recovered_r` becomes `kSigverifyShares` if it's correct and garbage
+  // otherwise.
+  //
+  // 2. Next, we produce `flash_exec` and `result`. `flash_exec` is produced by
+  // xor'ing all words of `recovered_r` with each other. If `recovered_r` is
+  // correct, `flash_exec` will be `kSigverifyFlashExec` due to the way
+  // `kSigverifyShares` is defined. To make sure that we don't produce this
+  // value otherwise, we compare each word of `recovered_r` with the
+  // corresponding expected value and set `flash_exec` to `UINT32_MAX` at each
+  // iteration if there is a mismatch. Finally, we produce the return value
+  // `result` from `flash_exec` by xor'ing parts of it together. Note that the
+  // hardware constant `kSigverifyFlashExec` is chosen such that this operation
+  // results in `kErrorOk`.
+
+  // Step 1: Process `recovered_r` so that it becomes `kSigverifyShares` if it's
+  // correct, garbage otherwise.
+  uint32_t *recovered_r_ptr = recovered_r->data;
+  size_t i = 0;
+  for (size_t j = 0; launder32(j) < kP256SignatureCordWords; ++j, ++i) {
+    recovered_r_ptr[i] ^= signature->data[j] ^ kSigverifyShares[i];
+  }
+  HARDENED_CHECK_EQ(i, kP256SignatureCordWords);
```

Comment:
are i and j both needed here for hardening? If so, should there be a hardened check on j too?

Reply:
I think it is ok to just harden `i` in this case. If `j` is attacked, the recovered value will get corrupted and result in error.